File: pubsdk-android-unity-wrapper/src/main/java/com/gamepubcorp/sdk/unitywrapper/PubSdkWrapper.kt
package com.gamepubcorp.sdk.unitywrapper

import io.github.gamepubcorp.api.PubApiClient
import io.github.gamepubcorp.api.PubApiClientBuilder
import com.gamepubcorp.sdk.unitywrapper.CallbackMessageForUnity.Companion.sendMessageError
import com.gamepubcorp.sdk.unitywrapper.auth.PubSdkWrapperActivity
import com.google.gson.Gson
import com.unity3d.player.UnityPlayer
import io.github.gamepubcorp.result.*
import io.github.gamepubcorp.utils.PubCallback


class PubSdkWrapper {

    private val TAG = "bridge"
    private lateinit var pubApiClient: PubApiClient
    private val gson = Gson()

    fun setupSDK(identifier: String,
                 projectId: String
    ) {
        val currentActivity = UnityPlayer.currentActivity

        pubApiClient = PubApiClientBuilder(currentActivity, projectId.toInt()).build()
        pubApiClient.setupSDK(currentActivity, PubCallback<PubSetupResult>().apply {
            success = { res ->
                val result = gson.toJson(res)
                CallbackMessageForUnity(identifier, result).sendMessageOk()
            }
            error = { err ->
                sendMessageError(identifier, err)
            }
        })
    }

    fun login(identifier: String,
              loginType: Int,
              accountServiceType: Int
    ) {
        val currentActivity = UnityPlayer.currentActivity
//        if (!this::pubApiClient.isInitialized) {
//            sendSdkNotInitializedError(currentActivity, identifier)
//            return
//        }
        PubSdkWrapperActivity.startActivity(
            currentActivity,
            identifier,
            loginType,
            accountServiceType
        )
    }

    fun autoLogin(identifier: String) {
        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.autologin(currentActivity, PubCallback<PubLoginResult>().apply {
            success = { res ->
                val result = gson.toJson(res)
                CallbackMessageForUnity(identifier, result).sendMessageOk()
            }
            error = { err ->
                sendMessageError(identifier, err)
            }
        })
    }

    fun logout(identifier: String) {

        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.logout(currentActivity, PubCallback<PubUnit>().apply {
            success = { res ->
                val result = gson.toJson(res)
                CallbackMessageForUnity(identifier, result).sendMessageOk()
            }
            error = { err ->
                sendMessageError(identifier, err)
            }
        })
    }

    fun withdraw(identifier: String) {

        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.withdraw(currentActivity, PubCallback<PubUnit>().apply {
            success = { res ->
                val result = gson.toJson(res)
                CallbackMessageForUnity(identifier, result).sendMessageOk()
            }
            error = { err ->
                sendMessageError(identifier, err)
            }
        })
    }

    fun initBilling(identifier: String) {

        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.initBilling(currentActivity, PubCallback<PubInitBillingResult>().apply {
            success = { res ->
                val result = gson.toJson(res)
                CallbackMessageForUnity(identifier, result).sendMessageOk()
            }
            error = { err ->
                sendMessageError(identifier, err)
            }
        })
    }

    fun purchase(identifier: String,
                 productId: String,
                 channelId: String,
                 characterId: String
    ) {
        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.purchase(
            currentActivity,
            productId,
            channelId,
            characterId,
            PubCallback<PubPurchaseResult>().apply {
                success = { res ->
                    val result = gson.toJson(res)
                    CallbackMessageForUnity(identifier, result).sendMessageOk()
                }
                error = { err ->
                    sendMessageError(identifier, err)
                }
            }
        )
    }

    fun retryPurchase(identifier: String,
                      channelId: String,
                      characterId: String
    ) {
        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.retryPurchase(
            currentActivity,
            channelId,
            characterId,
            PubCallback<PubRetryPurchaseResult>().apply {
                success = { res ->
                    val result = gson.toJson(res)
                    CallbackMessageForUnity(identifier, result).sendMessageOk()
                }
                error = { err ->
                    sendMessageError(identifier, err)
                }
            }
        )
    }

    fun openVoided(identifier: String,
                   channelId: String,
                   characterId: String
    ){
        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.openVoided(
            currentActivity,
            channelId,
            characterId, PubCallback<PubVoidedResult>().apply {
                success = { res ->
                    val result = gson.toJson(res)
                    CallbackMessageForUnity(identifier, result).sendMessageOk()
                }
                error = { err ->
                    sendMessageError(identifier, err)
                }
            }
        )
    }

    fun openTerms(identifier: String){

        val currentActivity = UnityPlayer.currentActivity

        pubApiClient.openTerms(currentActivity, PubCallback<PubUnit>().apply {
            success = { res ->
                val result = gson.toJson(res)
                CallbackMessageForUnity(identifier, result).sendMessageOk()
            }
            error = { err ->
                sendMessageError(identifier, err)
            }
        })
    }

    fun openImageBanner() {

        val currentActivity = UnityPlayer.currentActivity
        pubApiClient.openImageBanner(currentActivity)
    }

    fun openHelp() {

        val currentActivity = UnityPlayer.currentActivity
        pubApiClient.openHelp(currentActivity)
    }

    fun setPushToken(pushToken: String) {

        pubApiClient.setPushToken(pushToken)
    }

    fun setPushConfig(agreedPush: Boolean,
                      agreedNightPush: Boolean) {

        pubApiClient.setPushConfig(agreedPush, agreedNightPush)
    }
}

File: Pub-SDK-Unity/Assets/Plugins/iOS/PubSDK/PubSDKWrapper.m
//
//  PubSDKWrapper.m
//  PubSDKUnityBridge
//
//  Created by gamepub on 2022/05/16.
//

#import "PubSDKWrapper.h"
#import "PubSDKCallbackPayload.h"

@import PubSDKSwift;

@interface PubSDKWrapper()
@property (nonatomic, assign) BOOL setup;
@end

@implementation PubSDKWrapper

+ (instancetype)sharedInstance
{
    static PubSDKWrapper *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[PubSDKWrapper alloc] init];
    });
    return sharedInstance;
}

- (void)setupSDK:(NSString *)identifier
       projectId:(NSString *)projectId
{
    if(self.setup) {
        return;
    }
    self.setup = YES;
    
    NSLog(@"setupSDK");
    
    NSInteger projId = [projectId intValue];
        
    [[PubAPIClient shared] setupWithProjectId:projId
                            completionHandler:^(PubSDKSetupResult *result,
                                                NSError *error)
     {
        if(error) {
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[self wrapError:error]];
            [payload sendMessageError];
        }else{
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[result json]];
            [payload sendMessageOK];
        }
    }];
}

- (void)login:(NSString *)identifier
         type:(int)loginType
  serviceType:(int)accountServiceType
{
    NSLog(@"login");
    
    [[PubAPIClient shared] loginWithLoginType:loginType
                                  serviceType:accountServiceType
                             inViewController:UnityGetGLViewController()
                            completionHandler:^(PubSDKLoginResult *result,
                                                NSError *error)
     {
        if(error) {
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[self wrapError:error]];
            [payload sendMessageError];
        }else{
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[result json]];
            [payload sendMessageOK];
        }
    }];
}

- (void)autoLogin:(NSString *)identifier
{
    NSLog(@"autoLogin");
    
    [[PubAPIClient shared] autoLoginWithCompletionHandler:^(PubSDKLoginResult *result,
                                                            NSError *error)
     {
        if(error) {
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[self wrapError:error]];
            [payload sendMessageError];
        }else{
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[result json]];
            [payload sendMessageOK];
        }
    }];
}

- (void)logout:(NSString *)identifier
{
    NSLog(@"logout");
}

- (void)withdraw:(NSString *)identifier
{
    NSLog(@"withdraw");
    
    [[PubAPIClient shared] withdrawWithCompletionHandler:^(PubSDKLoginResult *result,
                                                           NSError *error)
     {
        if(error) {
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[self wrapError:error]];
            [payload sendMessageError];
        }else{
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[result json]];
            [payload sendMessageOK];
        }
    }];
}

- (void)initBilling:(NSString *)identifier
{
    NSLog(@"initBilling");
    
    [[PubAPIClient shared] initBillingWithCompletionHandler:^(PubSDKProductList *result,
                                                              NSError *error)
     {
        if(error) {
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[self wrapError:error]];
            [payload sendMessageError];
        }else{
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[result json]];
            [payload sendMessageOK];
        }
    }];
}

- (void)purchase:(NSString *)identifier
       productId:(NSString *)productId
       channelId:(NSString *)channelId
     characterId:(NSString *)characterId
{
    NSLog(@"purchase");
    
    [[PubAPIClient shared] purchaseWithProductId:productId
                                       channelId:channelId
                                     characterId:characterId
                               completionHandler:^(PubSDKPurchaseResult *result,
                                                   NSError *error)
     {
        if(error) {
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[self wrapError:error]];
            [payload sendMessageError];
        }else{
            PubSDKCallbackPayload *payload = [PubSDKCallbackPayload callbackMessage:identifier value:[result json]];
            [payload sendMessageOK];
        }
    }];
}

- (void)openTerms:(NSString *)identifier
{
    NSLog(@"openTerms");
}

- (void)openImageBanner
{
    NSLog(@"openImageBanner");
}

- (void)openHelp
{
    NSLog(@"openHelp");
}

- (void)setPushToken:(NSString *)pushToken
{
    NSLog(@"setPushToken");
    [[PubAPIClient shared] setPushTokenWithPushToken:pushToken];
}

- (void)setPushConfig:(BOOL)agreedPush
      agreedNightPush:(BOOL)agreedNightPush
{
    NSLog(@"setPushConfig");
}

- (NSString *)wrapError:(NSError *)error {
    NSDictionary *dic = @{@"code": @(error.code), @"message": error.localizedDescription};
    NSData *data = [NSJSONSerialization dataWithJSONObject:dic options:kNilOptions error:nil];
    if (!data) { return nil; }
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

@end
